pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,259 @@
+		return cfg
+	}
+
+	cfg[p.matchURL] = credentialprovider.DockerConfigEntry{
+		Username: response.AuthConfig.Username,
+		Password: response.AuthConfig.Password,
+		Email:    "plugin@criauth.k8s.io", // email is deprecated, this is just a dummy email
+	}
+
+	p.cachedDockerConfig = cfg
+	p.cacheExpiration = time.Now().Add(p.defaultCacheDuration)
+	if response.CacheDuration != nil {
+		if response.CacheDuration.Duration > 0 {
+			p.cacheExpiration = time.Now().Add(response.CacheDuration.Duration)
+		} else {
+			// expire cache immediately if cache duration in response was 0
+			p.cacheExpiration = time.Now()
+		}
+	}
+
+	return cfg
+}
+
+// Enabled always returns true since registration of the plugin via kubelet implies it should be enabled.
+func (e *pluginProvider) Enabled() bool {
+	return true
+}
+
+// IsPlugin returns true since all external credential providers are plugins
+func (e *pluginProvider) IsPlugin() bool {
+	return true
+}
+
+// Plugin is the interface calling ExecPlugin. This is mainly for testability
+// so tests don't have to actually exec any processes.
+type Plugin interface {
+	ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error)
+}
+
+// execPlugin is the implementation of the Plugin interface that execs a provided CRI auth plugin based
+// on it's name provided in CRIAuthPluginConfig. It is assumed that the executable is available in the
+// plugin directory provided by the kubelet.
+type execPlugin struct {
+	name         string
+	args         []string
+	pluginBinDir string
+}
+
+// ExecPlugin executes the plugin binary in the following format:
+//
+//  $ <plugin-name> get-credentials arg1=value1 arg2=value2
+//
+// The plugin is expected to receive the CRIAuthPluginRequest API via stdin and is expected
+// to return the CRIAuthPluginResponse via stdout.
+func (e *execPlugin) ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error) {
+	pluginBin := filepath.Join(e.pluginBinDir, e.name)
+	args := []string{"get-credentials"}
+	args = append(args, e.args...)
+
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	stdin := &bytes.Buffer{}
+
+	// Use a catch-all timeout of 1 minute for all exec-based plugins, this should leave enough
+	// head room in case a plugin needs to retry a failed request while ensuring an exec plugin
+	// does not run forever. In the future we may want this timeout to be tweakable from the plugin
+	// config file.
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, pluginBin, args...)
+	cmd.Stdout, cmd.Stderr, cmd.Stdin = stdout, stderr, stdin
+
+	authRequest := &criauth.CRIAuthPluginRequest{Image: image}
+	data, err := encodeRequest(authRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode auth request: %v", err)
+	}
+
+	if _, err := stdin.Write(data); err != nil {
+		return nil, fmt.Errorf("error writing to stdin: %v", err)
+	}
+
+	if err = cmd.Run(); err != nil {
```

Comment:
Should there maybe be a separate error message if the timeout occurs here? I was wondering how easy it was to differentiate between a plugin returning an error versus its process being cut off due to the timeout.

Reply:
I don't the wrapped error would include "context deadline exceeded" or something similar since `cmd` was created with `CommandContext`

Reply:
With this [small program](https://gist.github.com/DangerOnTheRanger/a85da22ed65f1cabf7ba753d50b48896) (I'm unfortunately not too familiar with `CommandContext` yet, so I went and copied this off the internet) I get the following output:
```
err is: signal: killed
context deadline exceeded
```

The first line in the output is the `err` from `cmd.Run()` after `cmd` was made with `CommandContext` and terminated due to passing the deadline, while the second is from printing `ctx.Err()`. So I think to get a more descriptive message, `ctx.Err()` will need to be checked additionally/separately from `cmd.Run()`. To help any SREs who might have to debug problems, it could also be helpful to return something like `plugin exceeded 1min timeout` (or something else grep-friendly indicating the container auth plugin timeout was exceeded) as the error message instead of what `CommandContext` provides by default, at least in the case of `context.DeadlineExceeded`.

Reply:
Great catch, i'll wrap the error here and log something more useful here

Reply:
Last commit checks for DeadlineExceeded https://github.com/kubernetes/kubernetes/pull/94196/commits/616556f8a0e88f9ca091efa9b5d92e9a96dc65ad -- please take a look :)